### ISSUE TITLE
Practice designing class hierarchies and relationships.

### ISSUE THREAD

**Issue by geekelo** [geekelo/geekelo#13]
Certainly! Designing class hierarchies and relationships is an important aspect of object-oriented programming. Let's go through a simple example to illustrate this concept. We'll design a class hierarchy for a zoo simulation.

### Zoo Simulation Example:

```javascript
// Animal class as the base class
class Animal {
  constructor(name, sound) {
    this.name = name;
    this.sound = sound;
  }

  makeSound() {
    console.log(`${this.name} says ${this.sound}`);
  }
}

// Mammal class extends Animal
class Mammal extends Animal {
  constructor(name, sound, furColor) {
    super(name, sound);
    this.furColor = furColor;
  }

  walk() {
    console.log(`${this.name} is walking.`);
  }
}

// Bird class extends Animal
class Bird extends Animal {
  constructor(name, sound, featherColor) {
    super(name, sound);
    this.featherColor = featherColor;
  }

  fly() {
    console.log(`${this.name} is flying.`);
  }
}

// Reptile class extends Animal
class Reptile extends Animal {
  constructor(name, sound, scaleType) {
    super(name, sound);
    this.scaleType = scaleType;
  }

  crawl() {
    console.log(`${this.name} is crawling.`);
  }
}

// Zoo class that aggregates different animals
class Zoo {
  constructor() {
    this.animals = [];
  }

  addAnimal(animal) {
    this.animals.push(animal);
  }

  performDailyActivities() {
    console.log("Daily activities at the zoo:");
    this.animals.forEach(animal => {
      if (animal instanceof Mammal) {
        animal.walk();
      } else if (animal instanceof Bird) {
        animal.fly();
      } else if (animal instanceof Reptile) {
        animal.crawl();
      }

      animal.makeSound();
    });
  }
}

// Usage
const lion = new Mammal("Lion", "Roar", "Golden");
const parrot = new Bird("Parrot", "Squawk", "Colorful");
const snake = new Reptile("Snake", "Hiss", "Smooth");

const zoo = new Zoo();
zoo.addAnimal(lion);
zoo.addAnimal(parrot);
zoo.addAnimal(snake);

zoo.performDailyActivities();
```

In this example:

- `Animal` is the base class with common properties and methods.
- `Mammal`, `Bird`, and `Reptile` extend the `Animal` class, inheriting its properties and methods while adding their specific attributes.
- `Zoo` is a class that aggregates instances of different animals and can perform daily activities for each type of animal.

This example demonstrates class hierarchies, inheritance, and polymorphism. Designing your class hierarchies thoughtfully can lead to more modular and maintainable code.

**Comment by geekelo** [comment]
You're correct, and I appreciate the opportunity to clarify. Abstraction involves exposing only the essential features of an object while hiding unnecessary details. It focuses on providing a clear and simplified interface to the outside world, while the internal implementation details are hidden.

In the context of programming, abstraction often involves defining interfaces, abstract classes, or protocols that specify the essential behaviors without exposing the underlying implementation. Clients using these abstractions don't need to know the intricacies of how things work internally; they only need to interact with the public interface.

In the zoo simulation example:

- The `Animal` class serves as an abstraction by defining a common interface (`makeSound()` method) that all animals share.
- The subclasses (`Mammal`, `Bird`, `Reptile`) further abstract specific details related to their respective types of animals by adding methods like `walk()`, `fly()`, and `crawl()`.

So, in summary, abstraction in programming involves exposing the essential features while hiding the implementation details, providing a simplified and understandable interface for users of the code.
